docs/decisions/0011-producing-to-event-bus-via-settings.rst

```diff
@@ -0,0 +1,106 @@
+11. Enable producing to event bus via settings
+##############################################
+
+Status
+******
+
+**Provisional** 2023-05-23
+
+Context
+*******
+
+Since the initial implementation of the event bus allowed only a single event
+type to be published to a topic, details like topic name, consumer group name,
+and consumer name were configured via code. The current implementation of
+openedx-events does not actually push any events to the underlying
+implementations like `edx-event-bus-kafka`_ and `edx-event-bus-redis`_. The
+event-producing application is expected to create a signal handler (since
+openedx-events subclasses Django signals) to catch the event and push it into
+the event bus. Some examples of the handlers can be found `here`_.
+
+.. _here: https://github.com/openedx/edx-platform/blob/27b8d2f68d5dfaf84755e7d7f8dccc97ce3be509/cms/djangoapps/contentstore/signals/handlers.py#L162-L210
+.. _edx-event-bus-kafka: https://github.com/openedx/event-bus-kafka
+.. _edx-event-bus-redis: https://github.com/openedx/event-bus-redis
+
+This ADR aims to propose a solution for configuring the details, like the topic.
+name, consumer group name, etc. via Django settings as well as pushing events to
+the event bus without requiring the producing application to write additional
+handlers.
+
+
+Decision
+********
+
+Create a generic signal handler to push events to the event bus. This handler
+should be attached to or connected to the signals that are enabled in Django settings.
+of the application. The configuration format will be as shown below:
+
+.. code-block:: python
+
+   OPEN_EDX_EVENTS_CONFIG = {
+       'org.openedx.content_authoring.xblock.deleted.v1': {
+           'topics': 'xblock-status',
+           'event_key_field': 'xblock_info.usage_key',
+           'event_data_key': 'xblock_info',
+           'enabled': True,
+       },
+       'org.openedx.content_authoring.xblock.published.v1': {
+           'topics': ['xblock-status', 'xblock-published'],
+           'event_key_field': 'xblock_info.usage_key',
+           'event_data_key': 'xblock_info',
+           'enabled': True,
+       }
+   }
+
+This configuration will be read in openedx_events ``apps.OpenedxEventsConfig(AppConfig).ready``
+method and a generic signal handler will be connected to the event_types (keys)
+listed in the configuration after validating its format.
+
+.. code-block:: python
```

Comment:
I don't know that there's much benefit to including implementation code in the ADR, as the actual code will diverge over time. It's probably sufficient to say that `openedx_events` will implement a Django `AppConfig.ready()` method to load the config and register signal handlers.

It might actually make sense to just have one PR that adds implementation and ADR at the same time, which we've done pretty often. I think the implementation approach should be uncontroversial enough that it would make sense to just go ahead and do that.

Reply:
@timmc-edx Makes sense, I wanted to include example of code to make it easier to understand. Updated it to just pseudo code now.